url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/evolution/engine/impl/ProblemToSTS.java

```diff
@@ -0,0 +1,759 @@
+			
+			DefaultLiteral traceLit = new DefaultLiteral("Trace" + index);
+			DefaultLiteral pointLit = new DefaultLiteral("Point" + index);
+			DefaultClause flagClause = new DefaultClause();
+			DefaultClause completeFlagClause = new DefaultClause();
+			
+			for (int i = 0; i < index; i++) {
+				DefaultLiteral flagLit = new DefaultLiteral("flag" + i);
+				completeFlagClause.addLiteral(flagLit);
+				if (i >= prevIndex) {
+					flagClause.addLiteral(flagLit);
+				}
+			}
+			Set<Clause<? extends Literal>> clauses = 
+				this.getRightPointClauses(traceLit, pointLit, flagClause);
+			
+			Transition triggerTrans = createTransition(initialState, 
+					condNoUncontrollable, trigger, nextState);
+			triggerTrans.addGuardClauses(clauses);
+			
+			Transition noTriggerTrans = createTransition(initialState, 
+					new StateFormula(), notrigger, nextState);
+			noTriggerTrans.addGuardClauses(clauses);
+			
+			Action reset = new DefaultAction("UNDEF", false, condition);
+			outputActions.add(reset);
+//			Transition resetTrans = createTransition(nextState, new StateFormula(), reset, initialState);
+//			resetTrans.addGuardClause(flagClause);
+			
+			Transition stayTrans = createTransition(nextState, new StateFormula(), reset, nextState);
+			for (int i = 0; i < index; i++) {
+				DefaultLiteral negFlagLit = new DefaultLiteral("flag" + i, true);
+				DefaultClause negFlagClause = new DefaultClause(negFlagLit);
+				stayTrans.addGuardClause(negFlagClause);
+			}
+			
+			Correction corr = corrections.get(condition.getIndex() - 1);
+			ProcessModel matchingAdaptationModel = corr.getAdaptation().getAdaptationModel();
+			
+//			Action adUndef = new DefaultAction("UNDEF", true, matchingAdaptationModel);	
+			
+			for (ProcessModel model : processModels) {
+				if (!model.equals(matchingAdaptationModel)) {
+					Action otherwiseAct = new WildcardAction(true, model);
+					inputActions.add(otherwiseAct);
+					Transition trans = createTransition(nextState, StateFormula.getTop(), otherwiseAct, initialState);
+					trans.addGuardClause(completeFlagClause);
+				}
+			}
+			
+			Transition resetTrans2 = createTransition(initialState, new StateFormula(), reset, initialState);
+			for (int i = prevIndex; i < index; i++) {
+				DefaultLiteral flagLit = new DefaultLiteral("flag" + i);
+				DefaultClause guardClause = 
+					new DefaultClause(traceLit.getNegation(), pointLit.getNegation(), flagLit.getNegation());
+				resetTrans2.addGuardClause(guardClause);
+			}
+			
+			this.updateDomainObjectTransitions(transToUpdate, trigger, clauses);
+		}
+		
+		return new DefaultSTS(states, initialState, inputActions, 
+				outputActions, transitions);		
+	}
+	
+	private void updateDomainObjectTransitions(
+			Map<ObjectTransition, DomainObject> transToUpdate, Action trigger, 
+			Set<Clause<? extends Literal>> rightPointClauses) {
+		
+		for (ObjectTransition trans : transToUpdate.keySet()) {
+			STS objSTS = objectSTSs.get(transToUpdate.get(trans));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.